File: src/ios/uoyabause/uoyabause/LeaderBoardViewModel.swift
import Foundation
import FirebaseAuth
import FirebaseFirestore

class LeaderBoardViewModel {
    // MARK: - Properties
    let repository = LeaderBoardRepository()
    private let pageSize = 100

    // Leaderboards list
    private(set) var leaderboards: [LeaderBoardRepository.LeaderboardInfo]? {
        didSet {
            onLeaderboardsChanged?(leaderboards ?? [])
        }
    }

    // Current leaderboard ID
    private(set) var currentLeaderboardId: String? {
        didSet {
            if oldValue != currentLeaderboardId, let leaderboardId = currentLeaderboardId {
                // Get top score for this leaderboard and save it
                Task {
                    do {
                        topScore = try await repository.getTopScore(leaderboardId: leaderboardId)
                        print("Set top score for leaderboard \(leaderboardId): \(topScore)")

                        // Get user position and load page
                        await loadUserPositionAndPage()
                    } catch {
                        print("Error getting top score: \(error)")
                        topScore = 0
                    }
                }
            }
        }
    }

    // User position
    private(set) var userPosition: LeaderBoardRepository.UserPosition? {
        didSet {
            onUserPositionChanged?(userPosition)
        }
    }

    // Total score count
    private(set) var totalScoreCount: Int = 0 {
        didSet {
            onTotalScoreCountChanged?(totalScoreCount)
        }
    }

    // User's rank
    private(set) var myRank: Int = 0 {
        didSet {
            onMyRankChanged?(myRank)
        }
    }

    // Scores list
    private(set) var scores: [LeaderBoardRepository.ScoreEntry]? {
        didSet {
            onScoresChanged?(scores ?? [])
        }
    }

    // Loading state
    private(set) var isLoading: Bool = false {
        didSet {
            onLoadingStateChanged?(isLoading)
        }
    }

    // Error message
    private(set) var error: String? {
        didSet {
            onErrorReceived?(error)
        }
    }

    // Close fragment flag
    private(set) var closeFragment: Bool = false {
        didSet {
            onCloseRequested?(closeFragment)
        }
    }

    // Scroll to my rank flag
    private(set) var scrollToMyRank: Bool = false {
        didSet {
            if scrollToMyRank {
                onScrollToMyRankRequested?(true)
            }
        }
    }

    // Pagination flags
    private(set) var hasMoreBefore: Bool = false
    private(set) var hasMoreAfter: Bool = false

    // Pagination documents
    private var firstVisible: DocumentSnapshot?
    private var lastVisible: DocumentSnapshot?

    // Rank tracking
    private var firstRank: Int = 0
    private var lastRank: Int = 0

    // Top score for the current leaderboard
    private var topScore: Int64 = 0

    // MARK: - Callbacks

    var onLeaderboardsChanged: (([LeaderBoardRepository.LeaderboardInfo]) -> Void)?
    var onScoresChanged: (([LeaderBoardRepository.ScoreEntry]) -> Void)?
    var onUserPositionChanged: ((LeaderBoardRepository.UserPosition?) -> Void)?
    var onLoadingStateChanged: ((Bool) -> Void)?
    var onErrorReceived: ((String?) -> Void)?
    var onCloseRequested: ((Bool) -> Void)?
    var onMyRankChanged: ((Int) -> Void)?
    var onTotalScoreCountChanged: ((Int) -> Void)?
    var onScrollToMyRankRequested: ((Bool) -> Void)?

    // MARK: - Public Methods

    // Load leaderboards for a game
    func loadLeaderboards(gameCode: String) {
        Task {
            await MainActor.run {
                isLoading = true
            }
            do {
                let result = try await repository.getLeaderboards(gameCode: gameCode)

                await MainActor.run {
                    self.leaderboards = result

                    if !result.isEmpty {
                        // Select first leaderboard
                        selectLeaderboard(leaderboardId: result[0].id)
                    } else {
                        // No leaderboards available
                        self.error = NSLocalizedString("Leaderboard not supported for this game", comment: "Leaderboard not supported error")
                        self.closeFragment = true
                    }
                }
            } catch {
                print("Error loading leaderboards: \(error)")
                await MainActor.run {
                    self.error = NSLocalizedString("Leaderboard not supported for this game", comment: "Leaderboard not supported error")
                    self.closeFragment = true
                }
            }
            await MainActor.run {
                isLoading = false
            }
        }
    }

    // Select a leaderboard
    func selectLeaderboard(leaderboardId: String) {
        if currentLeaderboardId == leaderboardId { return }

        currentLeaderboardId = leaderboardId
    }

    // Load user position and the page containing it
    func loadUserPositionAndPage() async {
        guard let leaderboardId = currentLeaderboardId else { return }
        let userId = Auth.auth().currentUser?.uid

        await MainActor.run {
            isLoading = true
        }

        do {
            if let userId = userId {
                // Get page with user's score
                let pageResult = try await repository.loadPage(
                    leaderboardId: leaderboardId,
                    pageNumber: 0,
                    userId: userId,
                    topScore: topScore
                )

                // Save results on the main thread
                await MainActor.run {
                    self.scores = pageResult.scores
                    self.firstVisible = pageResult.firstVisible
                    self.lastVisible = pageResult.lastVisible
                    self.hasMoreBefore = pageResult.hasMoreBefore
                    self.hasMoreAfter = pageResult.hasMoreAfter
                    self.totalScoreCount = pageResult.totalCount

                    // Update rank information
                    if let firstScore = pageResult.scores.first, let lastScore = pageResult.scores.last {
                        self.firstRank = firstScore.rank
                        self.lastRank = lastScore.rank
                    }

                    // Find user's score in the list
                    if let userScore = pageResult.scores.first(where: { $0.userId == userId }) {
                        self.userPosition = LeaderBoardRepository.UserPosition(
                            userId: userId,
                            rank: userScore.rank,
                            score: userScore.score
                        )
                        self.myRank = userScore.rank
                    }

                    // Set flag to scroll to user's position
                    self.scrollToMyRank = true
                }
            } else {
                // No user logged in, load first page
                await loadFirstPage()
            }
        } catch {
            print("Error loading user position: \(error)")
            await MainActor.run {
                self.error = NSLocalizedString("Failed to load leaderboard: \(error.localizedDescription)", comment: "Leaderboard loading error")
            }
        }

        await MainActor.run {
            isLoading = false
        }
    }

    // Load first page of scores
    func loadFirstPage() async {
        guard let leaderboardId = currentLeaderboardId else { return }

        await MainActor.run {
            isLoading = true
        }

        do {
            let pageResult = try await repository.loadPage(
                leaderboardId: leaderboardId,
                pageNumber: 0,
                userId: nil,
                topScore: topScore
            )

            await MainActor.run {
                self.scores = pageResult.scores
                self.firstVisible = pageResult.firstVisible
                self.lastVisible = pageResult.lastVisible
                self.hasMoreBefore = pageResult.hasMoreBefore
                self.hasMoreAfter = pageResult.hasMoreAfter
                self.totalScoreCount = pageResult.totalCount

                // Update rank information
                if let firstScore = pageResult.scores.first, let lastScore = pageResult.scores.last {
                    self.firstRank = firstScore.rank
                    self.lastRank = lastScore.rank
                }

                // Reset scroll flag
                self.scrollToMyRank = false
            }
        } catch {
            print("Error loading first page: \(error)")
            await MainActor.run {
                self.error = NSLocalizedString("Failed to load page: \(error.localizedDescription)", comment: "Page loading error")
            }
        }

        await MainActor.run {
            isLoading = false
        }
    }

    // Load next page of scores
    func loadNextPage() {
        guard let leaderboardId = currentLeaderboardId,
              let lastDoc = lastVisible,
              !isLoading,
              hasMoreAfter else { return }

        Task {
            await MainActor.run {
                isLoading = true
            }

            do {
                let pageResult = try await repository.loadScoresAfter(
                    leaderboardId: leaderboardId,
                    lastVisible: lastDoc,
                    lastRank: lastRank,
                    topScore: topScore
                )

                await MainActor.run {
                    // Add new scores to the current list
                    var currentScores = self.scores ?? []

                    if !pageResult.scores.isEmpty {
                        // Update ranks for the next page
                        let nextPageScores = pageResult.scores.enumerated().map { index, score in
                            return LeaderBoardRepository.ScoreEntry(
                                userId: score.userId,
                                name: score.name,
                                score: score.score,
                                diff: score.diff,
                                timestamp: score.timestamp,
                                rank: lastRank + index + 1,
                                photoURL: score.photoURL
                            )
                        }

                        currentScores.append(contentsOf: nextPageScores)
                        self.scores = currentScores

                        // Update last rank
                        self.lastRank = self.lastRank + pageResult.scores.count
                    }

                    // Update last visible document
                    if let lastVisible = pageResult.lastVisible {
                        self.lastVisible = lastVisible
                    }

                    // Update pagination flags
                    self.hasMoreAfter = pageResult.hasMoreAfter
                }
            } catch {
                print("Error loading next page: \(error)")
                await MainActor.run {
                    self.error = NSLocalizedString("Failed to load more scores: \(error.localizedDescription)", comment: "Load more error")
                }
            }

            await MainActor.run {
                isLoading = false
            }
        }
    }

    // Load previous page of scores
    func loadPreviousPage() {
        guard let leaderboardId = currentLeaderboardId,
              let firstDoc = firstVisible,
              !isLoading,
              hasMoreBefore else { return }

        Task {
            await MainActor.run {
                isLoading = true
            }

            do {
                let pageResult = try await repository.loadScoresBefore(
                    leaderboardId: leaderboardId,
                    firstVisible: firstDoc,
                    firstRank: firstRank,
                    topScore: topScore
                )

                await MainActor.run {
                    // Add new scores to the beginning of the current list
                    var currentScores = self.scores ?? []

                    if !pageResult.scores.isEmpty {
                        // Update first rank
                        self.firstRank = pageResult.scores.first?.rank ?? self.firstRank

                        // Insert scores at the beginning
                        currentScores.insert(contentsOf: pageResult.scores, at: 0)
                        self.scores = currentScores
                    }

                    // Update first visible document
                    if let firstVisible = pageResult.firstVisible {
                        self.firstVisible = firstVisible
                    }

                    // Update pagination flags
                    self.hasMoreBefore = pageResult.hasMoreBefore
                }
            } catch {
                print("Error loading previous page: \(error)")
                await MainActor.run {
                    self.error = NSLocalizedString("Failed to load previous scores: \(error.localizedDescription)", comment: "Load previous error")
                }
            }

            await MainActor.run {
                isLoading = false
            }
        }
    }

    // MARK: - Helper Methods

    func clearError() {
        error = nil
    }

    func resetCloseFragment() {
        closeFragment = false
    }

    func resetScrollFlag() {
        scrollToMyRank = false
    }
}

File: src/android/app/src/main/java/org/uoyabause/android/LeaderBoardViewModel.kt
package org.uoyabause.android

import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.DocumentSnapshot
import kotlinx.coroutines.launch
import org.devmiyax.yabasanshiro.R

class LeaderBoardViewModel : ViewModel() {
    var repository = LeaderBoardRepository()
    private val pageSize = 100

    companion object {
        const val TAG = "LeaderBoardViewModel"
    }

    // リーダーボード一覧
    private val _leaderboards = MutableLiveData<List<LeaderBoardRepository.LeaderboardInfo>>()
    val leaderboards: LiveData<List<LeaderBoardRepository.LeaderboardInfo>> = _leaderboards

    // 現在のリーダーボードID
    private val _currentLeaderboardId = MutableLiveData<String>()
    val currentLeaderboardId: LiveData<String> = _currentLeaderboardId

    // 現在のユーザー位置情報
    private val _userPosition = MutableLiveData<LeaderBoardRepository.UserPosition?>()
    val userPosition: LiveData<LeaderBoardRepository.UserPosition?> = _userPosition

    // 総スコア数
    private val _totalScoreCount = MutableLiveData<Int>(0)
    val totalScoreCount: LiveData<Int> = _totalScoreCount

    // 自分の順位
    private val _myRank = MutableLiveData<Int>(0)
    val myRank: LiveData<Int> = _myRank

    // 現在表示中のスコアリスト
    private val _scores = MutableLiveData<List<LeaderBoardRepository.ScoreEntry>>()
    val scores: LiveData<List<LeaderBoardRepository.ScoreEntry>> = _scores

    // 前後のページがあるかどうか
    private val _hasMoreBefore = MutableLiveData<Boolean>(false)
    val hasMoreBefore: LiveData<Boolean> = _hasMoreBefore

    private val _hasMoreAfter = MutableLiveData<Boolean>(false)
    val hasMoreAfter: LiveData<Boolean> = _hasMoreAfter

    // 最初と最後のドキュメント（ページネーション用）
    private var firstVisible: DocumentSnapshot? = null
    private var lastVisible: DocumentSnapshot? = null

    // 最初と最後の順位情報
    private var firstRank: Int = 0
    private var lastRank: Int = 0
    private var topScore: Long = 0

    // ローディング状態
    private val _isLoading = MutableLiveData<Boolean>(false)
    val isLoading: LiveData<Boolean> = _isLoading

    // エラー状態
    private val _error = MutableLiveData<String?>()
    val error: LiveData<String?> = _error

    // 自分の順位にスクロールする必要があるかどうか
    private val _scrollToMyRank = MutableLiveData<Boolean>(false)
    val scrollToMyRank: LiveData<Boolean> = _scrollToMyRank

    // 前のページの最後にスクロールする必要があるかどうか
    private val _scrollToPreviousPageBottom = MutableLiveData<Int>(-1)
    val scrollToPreviousPageBottom: LiveData<Int> = _scrollToPreviousPageBottom

    // 前のページのスクロールフラグをリセット
    fun resetScrollToPreviousPageBottom() {
        _scrollToPreviousPageBottom.value = -1
    }

    // フラグメントを閉じるためのイベント
    private val _closeFragment = MutableLiveData<Boolean>(false)
    val closeFragment: LiveData<Boolean> = _closeFragment

    // フラグメントを閉じるイベントをリセット
    fun resetCloseFragment() {
        _closeFragment.value = false
    }

    // リーダーボード一覧を読み込む
    fun loadLeaderboards(gameCode: String) {
        viewModelScope.launch {
            _isLoading.value = true
            try {
                val result = repository.getLeaderboards(gameCode)
                _leaderboards.value = result

                if (result.isNotEmpty()) {
                    // 最初のリーダーボードを選択
                    selectLeaderboard(result[0].id)
                } else {
                    // リーダーボードが存在しない場合
                    _error.value = YabauseApplication.appContext.getString(R.string.leaderboard_not_supported)
                    _closeFragment.value = true
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error loading leaderboards", e)
                _error.value = YabauseApplication.appContext.getString(R.string.leaderboard_not_supported)
                _closeFragment.value = true
                _isLoading.value = false
            } finally {
            }
        }
    }

    // リーダーボードを選択
    fun selectLeaderboard(leaderboardId: String) {
        if (_currentLeaderboardId.value == leaderboardId) return

        _currentLeaderboardId.value = leaderboardId

        // このリーダーボードのトップスコアを取得して保存
        viewModelScope.launch {
            try {
                topScore = repository.getTopScore(leaderboardId)
                Log.d(TAG, "Set top score for leaderboard $leaderboardId: $topScore")

                // ユーザーがログインしているかどうかを確認
                val userId = FirebaseAuth.getInstance().currentUser?.uid
                if (userId != null) {
                    // ログインしている場合は自分の順位を取得して表示
                    loadUserPositionAndPage()
                } else {
                    // ログインしていない場合は最初のページを表示
                    // topScoreは既に設定されているので、diffは正しく計算される
                    loadFirstPage()
                }

            } catch (e: Exception) {
                Log.e(TAG, "Error getting top score", e)
                topScore = 0L
                // エラーが発生した場合でも最初のページを表示
                loadFirstPage()
            }
        }
    }

    // 現在のユーザー位置を取得し、そのページを読み込む
    fun loadUserPositionAndPage() {
        val leaderboardId = _currentLeaderboardId.value ?: return
        val userId = FirebaseAuth.getInstance().currentUser?.uid

        viewModelScope.launch {
            _isLoading.value = true
            try {
                if (userId != null) {
                    // ユーザーIDを指定してページを取得（ユーザーのスコアが含まれるページが自動的に取得される）
                    val pageResult = repository.loadPage(leaderboardId, 0, userId,topScore)

                    // 結果を保存
                    _scores.value = pageResult.scores
                    firstVisible = pageResult.firstVisible
                    lastVisible = pageResult.lastVisible
                    _hasMoreBefore.value = pageResult.hasMoreBefore
                    _hasMoreAfter.value = pageResult.hasMoreAfter

                    // 順位情報を更新
                    if (pageResult.scores.isNotEmpty()) {
                        firstRank = pageResult.scores.first().rank
                        lastRank = pageResult.scores.last().rank
                    }

                    // ユーザーの順位を取得して設定
                val userScore = pageResult.scores.find { it.userId == userId }

                // 総数を取得
                val totalCount = repository.getTotalScoreCount(leaderboardId)
                _totalScoreCount.value = totalCount

                if (userScore != null) {
                    // 自分の順位を設定
                    _myRank.value = userScore.rank

                    _userPosition.value = LeaderBoardRepository.UserPosition(
                        userId = userId,
                        score = userScore.score,
                        rank = userScore.rank,
                        totalCount = totalCount,
                        pageNumber = 0
                    )

                    // ユーザーの順位が中央になるようにスクロール
                    _scrollToMyRank.value = true
                } else {
                    // ユーザーのスコアがページに含まれていない場合は、ユーザー位置をクリア
                    _userPosition.value = null
                    _myRank.value = 0
                }
                } else {
                    // ユーザーIDがない場合は最初のページを表示
                    loadFirstPage()
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error loading user position and page", e)
                _error.value = "ユーザー位置の読み込みに失敗しました: ${e.message}"
               // loadFirstPage() // エラー時は最初のページを表示
            } finally {
                _isLoading.value = false
            }
        }
    }

    // 最初のページを読み込む
    fun loadFirstPage() {
        val leaderboardId = _currentLeaderboardId.value ?: return

        viewModelScope.launch {
            _isLoading.value = true
            try {
                val pageResult = repository.loadPage(leaderboardId, 0, null, topScore)

                _scores.value = pageResult.scores
                firstVisible = pageResult.firstVisible
                lastVisible = pageResult.lastVisible
                _hasMoreBefore.value = pageResult.hasMoreBefore
                _hasMoreAfter.value = pageResult.hasMoreAfter

                // 順位情報を更新
                if (pageResult.scores.isNotEmpty()) {
                    firstRank = pageResult.scores.first().rank
                    lastRank = pageResult.scores.last().rank
                }

                // スクロールフラグをリセット
                _scrollToMyRank.value = false
            } catch (e: Exception) {
                Log.e(TAG, "Error loading first page", e)
                _error.value = "ページの読み込みに失敗しました: ${e.message}"
            } finally {
                _isLoading.value = false
            }
        }
    }

    // 次のページを読み込む
    fun loadNextPage() {
        val leaderboardId = _currentLeaderboardId.value ?: return
        val lastDoc = lastVisible ?: return

        viewModelScope.launch {
            _isLoading.value = true
            try {
                val pageResult = repository.loadScoreAfter(leaderboardId, lastDoc, lastRank, topScore)

                // 現在のスコアリストに新しいスコアを追加
                val currentScores = _scores.value ?: emptyList()

                // 順位情報を更新してからスコアを追加
                if (pageResult.scores.isNotEmpty()) {
                    // 次のページの順位は前のページの最後の順位から続くようにする
                    val nextPageScores = pageResult.scores.mapIndexed { index, score ->
                        score.copy(rank = lastRank + index + 1)
                    }
                    _scores.value = currentScores + nextPageScores

                    // 最後の順位を更新
                    lastRank = lastRank + pageResult.scores.size
                } else {
                    _scores.value = currentScores + pageResult.scores
                }

                // 最後のドキュメントを更新
                if (pageResult.lastVisible != null) {
                    lastVisible = pageResult.lastVisible
                }

                // 前後のページがあるかどうかを更新
                _hasMoreBefore.value = true // 次のページを読み込んだので前のページは必ずある
                _hasMoreAfter.value = pageResult.hasMoreAfter
            } catch (e: Exception) {
                Log.e(TAG, "Error loading next page", e)
                _error.value = "次のページの読み込みに失敗しました: ${e.message}"
            } finally {
                _isLoading.value = false
            }
        }
    }

    // 前のページを読み込む
    fun loadPreviousPage() {
        val leaderboardId = _currentLeaderboardId.value ?: return
        val firstDoc = firstVisible ?: return

        viewModelScope.launch {
            _isLoading.value = true
            try {
                val pageResult = repository.loadScoresBefore(leaderboardId, firstDoc, firstRank, topScore)

                // 現在のスコアリストに新しいスコアを追加（前に追加）
                val currentScores = _scores.value ?: emptyList()

                // 順位情報を更新
                if (pageResult.scores.isNotEmpty()) {
                    // 前のページの順位を更新
                    firstRank = pageResult.scores.first().rank
                    _scores.value = pageResult.scores + currentScores

                    // 前のページの最後にスクロールするようにフラグを設定
                    _scrollToPreviousPageBottom.value = pageResult.scores.size - 1
                } else {
                    _scores.value = pageResult.scores + currentScores
                }

                // 最初のドキュメントを更新
                if (pageResult.firstVisible != null) {
                    firstVisible = pageResult.firstVisible
                }

                // 前後のページがあるかどうかを更新
                _hasMoreBefore.value = pageResult.hasMoreBefore
                _hasMoreAfter.value = true // 前のページを読み込んだので次のページは必ずある
            } catch (e: Exception) {
                Log.e(TAG, "Error loading previous page", e)
                _error.value = "前のページの読み込みに失敗しました: ${e.message}"
            } finally {
                _isLoading.value = false
            }
        }
    }

    // スクロールフラグをリセット
    fun resetScrollFlag() {
        _scrollToMyRank.value = false
    }

    // エラーメッセージをクリア
    fun clearError() {
        _error.value = null
    }
}
